go/webpack/cbor_format.go

```diff
@@ -35,42 +35,92 @@ func parseIndexedContent(reader *bytes.Reader, parts []*PackPart) error {
 	panic("Not implemented")
 }
 
-func WriteCbor(p *Package) (result []byte, err error) {
-	cborPackage := cbor.New()
+func WriteCbor(p *Package, to io.Writer) error {
+	// Write the indexed-content/responses array first in order to compute
+	// the offsets of each response within it.
+	responsesFile, err := ioutil.TempFile("", "webpack-responses")
+	if err != nil {
+		return err
+	}
+	defer responsesFile.Close()
+	defer os.Remove(responsesFile.Name())
+
+	partOffsets, err := writeCborResourceBodies(p, responsesFile)
+	if err != nil {
+		return err
+	}
+
+	cborPackage := cbor.New(to)
 
 	arr := cborPackage.AppendArray(5)
 
 	// "🌐📦" in UTF-8.
 	var magicNumber = []byte{0xF0, 0x9F, 0x8C, 0x90, 0xF0, 0x9F, 0x93, 0xA6}
 
 	arr.AppendBytes(magicNumber)
+
+	// section-offsets:
 	sectionOffsets := arr.AppendMap(1)
 	sectionOffsets.AppendUtf8S("indexed-content")
 	// "indexed-content" will appear at the start of the 'sections' map.
-	indexedContentOffset := sectionOffsets.AppendPendingUint()
+	const indexedContentOffset = 1
+	sectionOffsets.AppendUint64(indexedContentOffset)
 	sectionOffsets.Finish()
 
 	sections := arr.AppendMap(1)
-	indexedContentOffset.Complete(uint64(sections.ByteLenSoFar()))
+
+	// indexed-content major section:
+	if sections.ByteLenSoFar() != indexedContentOffset {
+		panic(fmt.Sprintf("Wrote incorrect offset (%v) for indexed-content section actually at offset %v",
+			indexedContentOffset, sections.ByteLenSoFar()))
+	}
 	sections.AppendUtf8S("indexed-content")
 	indexedContent := sections.AppendArray(2)
-	index := indexedContent.AppendArray(len(p.parts))
-	pendingOffsets := make(map[*PackPart]*cbor.PendingInt, len(p.parts))
+
+	// Write the requests and the byte offsets to their responses into the
+	// index.
+	index := indexedContent.AppendArray(uint64(len(p.parts)))
 	for _, part := range p.parts {
 		arr := index.AppendArray(2)
 		arr.AppendBytes(encodeResourceKey(part))
-		pendingOffsets[part] = arr.AppendPendingUint()
+		partOffset, ok := partOffsets[part]
+		if !ok {
+			panic(fmt.Sprintf("%p missing from %v", part, partOffsets))
+		}
+		arr.AppendUint64(partOffset)
 		arr.Finish()
 	}
 	index.Finish()
-	responses := indexedContent.AppendArray(len(p.parts))
-	for _, part := range p.parts {
-		pendingOffset, ok := pendingOffsets[part]
-		if !ok {
-			panic(fmt.Sprintf("%p missing from %v", part, pendingOffsets))
+
+	{
```

Comment:
It seemed easier to read. Maybe it's not because it makes readers wonder why there's a block?